### DISCUSSION TITLE
Distributed Systems

### DISCUSSION THREAD

**Discussion by nlnpraveen** [JonPSmith/AuthPermissions.AspNetCore#30] [General]
Hi Jon,

Can the AuthP library be used to work with a sharded database to support a multi tenant application in a distributed system context? Meaning can we shard by region/geolocation and would there be any issues with latency and replication? If there is any documentation that be shared it will be helpful. Thanks.

-Praveen

**Comment by JonPSmith** [comment]
Hi @nlnpraveen,

The answer is AuthP is designed to support multiple database servers that - see the "The assumptions / rules" section of issue #29. Version 3.2.0 (which I am working on now) works by defining different SQL Servers via the ConnectionStrings section of the appsettings.json file and another file called shardingsettings.json defines the databases linked to a connection string in the ConnectionStrings section. This means you can have SQL servers geographically located which mean you can select the right SQL server / database for a tenant based on their location

Having said that building a multi-tenant app with tenants that are geographically spread is a challenge - I know because I worked on one. The main problem is if a tenant is spread over a large geographic area, then you need to do something more complex - like shared databases or using Cosmos DB which is easier to share (AuthP can support Cosmos DB shards - you just have to write some extra code). That's because AuthP (any other libraries) associate a tenant to **one** database. Also you will need geographically spread web servers and ways to send the traffic to nearest / correct web server. Its a big job!
